src/main/java/it/gov/pagopa/pu/fileshare/service/FileService.java

```diff
@@ -1,18 +1,98 @@
 import org.springframework.stereotype.Service;
 import org.springframework.web.multipart.MultipartFile;
 
 @Slf4j
 @Service
 public class FileService {
+  private final String sharedFolderRootPath;
+  private final String fileEncryptPassword;
+
+  public FileService(@Value("${folders.shared}") String sharedFolderRootPath,
+    @Value(("${app.fileEncryptPassword}")) String fileEncryptPassword) {
+    this.sharedFolderRootPath = sharedFolderRootPath;
+    this.fileEncryptPassword = fileEncryptPassword;
+  }
+
   public void validateFile(MultipartFile ingestionFlowFile, String validFileExt) {
-    if( ingestionFlowFile == null || !StringUtils.defaultString(ingestionFlowFile.getOriginalFilename()).endsWith(validFileExt)){
+    if( ingestionFlowFile == null){
+      log.debug("Invalid ingestion flow file");
+      throw new InvalidFileException("Invalid file");
+    }
+    String filename = StringUtils.defaultString(ingestionFlowFile.getOriginalFilename());
+    validateFileExtension(validFileExt, filename);
+    validateFilename(filename);
+  }
+
+  private static void validateFilename(String filename) {
+    if(Stream.of("..", "\\", "/").anyMatch(filename::contains)){
+      log.debug("Invalid ingestion flow filename");
+      throw new InvalidFileException("Invalid filename");
+    }
+  }
+
+  private static void validateFileExtension(String validFileExt, String filename) {
+    if(!filename.endsWith(validFileExt)){
       log.debug("Invalid ingestion flow file extension");
       throw new InvalidFileException("Invalid file extension");
     }
   }
+
+  private Path getFilePath(String relativePath, String filename) {
+    String basePath = sharedFolderRootPath+relativePath;
+    Path fileLocation = Paths.get(basePath,filename).normalize();
+    if(!fileLocation.startsWith(basePath)){
+      log.debug("Invalid file path");
+      throw new InvalidFileException("Invalid file path");
+    }
+    return fileLocation;
+  }
+
+  public void saveToSharedFolder(MultipartFile file, String relativePath){
+    if(file==null){
+      log.debug("File is mandatory");
+      throw new FileUploadException("File is mandatory");
+    }
+
+    String filename = org.springframework.util.StringUtils.cleanPath(StringUtils.defaultString(file.getOriginalFilename()));
+    validateFilename(filename);
+    Path fileLocation = getFilePath(relativePath, filename);
+    //create missing parent folder, if any
+    try {
+      if (!fileLocation.toAbsolutePath().getParent().toFile().exists())
+        Files.createDirectories(fileLocation.toAbsolutePath().getParent());
+      encryptAndSaveFile(file, fileLocation);
+    }catch (Exception e) {
+      log.debug(
+        "Error uploading file to folder %s%s".formatted(sharedFolderRootPath,
+          relativePath), e);
+      throw new FileUploadException(
+        "Error uploading file to folder %s%s [%s]".formatted(
+          sharedFolderRootPath, relativePath, e.getMessage()));
```

Comment:
don't duplicate message error, just build a String ad re-use it

ps: exception handling would also print the message?